src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,335 @@
+    U_b, s, Vh_b = sp.linalg.svd(R_B.T, full_matrices=False) 
+
+    with logger.block(f'Computing generalized left-singular vectors ({modes} vectors) ...'):
+        U = Q.lincomb(U_b.T)
+
+    with logger.block(f'Computung generalized right-singular vector ({modes} vectors) ...'):
+        Vh = Q_B.lincomb(Vh_b)
+        
+    return  U[:modes], s[:modes], Vh[:modes]
+
+
+@defaults('modes', 'p', 'q')
+def random_ghep(A, E=None, modes=6, p=20, q=2, single_pass=False):
+    r"""Approximates a few eigenvalues of a symmetric linear |Operator| with randomized methods.
+
+    Approximates `modes` eigenvalues `w` with corresponding eigenvectors `v` which solve
+    the eigenvalue problem.
+    
+    .. math::
+        A v_i = w_i v_i
+
+    or the generalized eigenvalue problem
+
+    .. math::
+        A v_i = w_i E v_i
+
+    if `E` is not `None`.
+
+    This method is an implementation of algorithm 6 and 7 in :cite:`SJK21`. 
+
+    Parameters
+    ----------
+    A
+        The hermitian linear |Operator| for which the eigenvalues are to be computed.
+    E
+        The hermitian |Operator| which defines the generalized eigenvalue problem.
+    modes
+        The number of eigenvalues and eigenvectors which are to be computed.
+    p :
+        If not `0`, adds `p` colums to the randomly sampled matrix in the :func:`rrf` method 
+        (oversampling parameter).
+    q : 
+        If not `0`, performs `q` power iterations to increase the relative weight
+        of the larger singular values.
+    single_pass
+        If `True`, computes the ghep where only one set of matvec Ax is required, but an additional error is produced. 
+        If `False` the methods requires two sets of matvecs Ax
+
+    Returns
+    -------
+    w
+        A 1D |NumPy array| which contains the computed eigenvalues.
+    V
+        A |VectorArray| which contains the computed eigenvectors.
+    """
+    logger = getLogger('pymor.algorithms.rand_la')
+
+    assert isinstance(A, Operator) and A.linear
+    assert not A.parametric
+    assert A.source == A.range
+    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes, int)
+    assert 0 <= p <= max(A.source.dim, A.range.dim) - modes and isinstance(p, int)
+    assert q >= 0 and isinstance(q, int)
+    
+    if E is None:
+        E = IdentityOperator(A.source)
+    else:
+        assert isinstance(E, Operator) and E.linear
+        assert not E.parametric
+        assert E.source == E.range
+        assert E.source == A.source
+    
+    if A.source.dim == 0 or A.range.dim == 0:
+        return A.source.empty(), np.array([]), A.range.empty()
+    
+    if single_pass:
+        Omega = A.source.random(modes+p, distribution='normal')
+        Y_bar = A.apply(Omega)
+        Y = InverseOperator(E).apply(Y_bar)
+        Q, R = gram_schmidt(Y, product=E, return_R=True)
+        X = E.apply2(Omega,Q)
+        Z = E.apply2(Q, Omega)
+        T = np.linalg.inv(X) @ Omega.inner(Y_bar) @ np.linalg.inv(Z)
```

Comment:
Then you would also only have to call `inv` once since the transpose of the inverse is the inverse of the transpoe.